modules/core/src/main/java/org/apache/ignite/internal/util/future/GridFutureAdapter.java

```diff
@@ -379,6 +379,28 @@ private void unblock(Object waiter) {
         return fut;
     }
 
+    /** {@inheritDoc} */
+    @Override public <T> IgniteInternalFuture<T> chainCompose(
+        IgniteClosure<? super IgniteInternalFuture<R>, IgniteInternalFuture<T>> doneCb
+    ) {
+        GridFutureAdapter<T> res = new GridFutureAdapter<>();
+
+        listen(fut -> {
+            IgniteInternalFuture<T> doneCbFut = doneCb.apply(fut);
```

Comment:
This call can be a cause of `Exception` I think. In that case `res.onDone()` will not be called at all, which is a problem I think.
Please take a look at `GridFutureChainListener.applyCallback` for a reference implementation.